## Constantes

In [76]:
import numpy as np


me =  9.1094 * 10**(-31)
e =  1.602 * 10**(-19)
hbar = 1.056 * 10**(-34)
pi = np.pi

## Question 4

Tout d'abord, on sait que n'importe quel état $\psi(x)$ peut être représenté en série de Fourier selon la forme suivante :
$$\psi(x) = \sum_{n=1}^{\infty} \psi_n\sin\frac{n\pi x}{L}$$
En insérant l'état $\psi(x)$ sous cette forme dans l'équation de Schrödinger, on obtient
$$\hat{H}\psi(x) = E\psi(x)$$
$$\hat{H}\sum_{n=1}^{\infty} \psi_n\sin\frac{n\pi x}{L} = E\sum_{n=1}^{\infty} \psi_n\sin\frac{n\pi x}{L}$$
Par la distributivité de $\hat{H}$ et comme les $\psi_n$ sont des constantes, on a
$$\sum_{n=1}^{\infty} \psi_n\hat{H}\sin\frac{n\pi x}{L} = E\sum_{n=1}^{\infty} \psi_n\sin\frac{n\pi x}{L}$$
On multiplie $\sin\frac{m\pi x}{L}$ par les deux côtés de l'équation et on distribue sur les sommes :
$$\sum_{n=1}^{\infty} \psi_n\sin\frac{m\pi x}{L}\hat{H}\sin\frac{n\pi x}{L} = E\sum_{n=1}^{\infty} \psi_n\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}$$
On intègre de 0 à $L$ des deux côtés de l'équation:
$$\int_0^L\sum_{n=1}^{\infty} \psi_n\sin\frac{m\pi x}{L}\hat{H}\sin\frac{n\pi x}{L}\text{d}x = \int_0^LE\sum_{n=1}^{\infty} \psi_n\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x$$
$$\sum_{n=1}^{\infty} \psi_n\int_0^L\sin\frac{m\pi x}{L}\hat{H}\sin\frac{n\pi x}{L}\text{d}x = E\sum_{n=1}^{\infty} \psi_n\int_0^L\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x$$
Or, on sait que
$$\int_0^L \sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x = 0 \qquad \text{pour $m\neq n$}$$
$$\int_0^L \sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x = \frac{L}{2} \qquad \text{pour $m=n$}$$
Tous les termes de la somme de droite seront donc nuls excepté celui où $n=m$. On a donc
$$\sum_{n=1}^{\infty} \psi_n\int_0^L\sin\frac{m\pi x}{L}\hat{H}\sin\frac{n\pi x}{L}\text{d}x = \frac{1}{2}LE\psi_m$$

En observant la forme de l'équation de Schrödinger obtenue pour la série de Fourier d'un état, on décide de poser des éléments de matrice $H_{mn}$ de la façon suivante :
$$H_{mn} = \frac{2}{L}\int_0^L\sin\frac{m\pi x}{L}\hat{H}\sin\frac{n\pi x}{L}\text{d}x$$

Cette définition nous permet de réécrire l'éqauation de Schrödinger sous une forme matricielle. En insérant $H_{mn}$ dans l'équation dérivée plus haut, on a
$$\sum_{n=1}^{\infty} \psi_n\frac{L}{2}H_{mn} = \frac{1}{2}LE\psi_m$$
$$\sum_{n=1}^{\infty} H_{mn} \psi_n = E\psi_m$$
On remarque que la partie gauche de l'équaiton correspond à une étape de la mutliplication matricielle de $\boldsymbol{H}$ et de $\boldsymbol{\psi}$ où on trouve l'élément $\psi_m$ de la matrice résultante. On remarque aussi que chaque élément de la matrice résultante est simplement l'élément de la matrice $\psi(x)$ multiplié par le facteur $E$. On voit donc que l'équation trouvée implique l'équation aux valeurs propres suivante :
$$\boldsymbol{H}\boldsymbol{\psi} = E\boldsymbol{\psi}$$
En définissant les éléments de la matrice $H$ de cette façon, on obtient donc que l'état $\psi(x)$ est un vecteur propre de l'Hamiltonien associé à la valeur propre E.

$$\textcolor{red}{\text{Ajouter du stock? Bonjour}}

## Question 5

Le potentiel est donné par
$$V(x) = \frac{ax}{L}$$
Ce qui donne l'Hamiltonien suivant :
$$\hat{H} = \frac{-\hbar^2}{2M}\frac{\text{d}^2}{\text{d}x^2} + \frac{ax}{L}$$
Les éléments de de matrice $H_{mn}$ sont donc donnés par :
$$H_{mn} = \frac{2}{L}\int_0^L\sin\frac{m\pi x}{L}\bigg(\frac{-\hbar^2}{2M}\frac{\text{d}^2}{\text{d}x^2} + \frac{ax}{L}\bigg)\sin\frac{n\pi x}{L}\text{d}x$$
$$H_{mn} = \frac{2}{L}\int_0^L\sin\frac{m\pi x}{L}\bigg[\frac{-\hbar^2}{2M}\bigg(-\frac{\pi^2n^2}{L^2}\sin\frac{n\pi x}{L}\bigg) + \frac{ax}{L}\sin\frac{n\pi x}{L}\bigg]\text{d}x$$
$$H_{mn} = \frac{2}{L^2}\int_0^L\sin\frac{m\pi x}{L}\bigg[\frac{\hbar^2\pi^2n^2}{2ML}\sin\frac{n\pi x}{L} + ax\sin\frac{n\pi x}{L}\bigg]\text{d}x$$
$$H_{mn} = \frac{\hbar^2\pi^2n^2}{ML^3}\int_0^L\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x + \frac{2a}{L^2}\int_0^Lx\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x$$
Or, en plus des propriétés énoncées plus haut, on sait également que
$$\int_0^Lx\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x = 0 \qquad \text{pour $m\neq n$ et $m+n$ est pair}$$
$$\int_0^Lx\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x = -\frac{4L^2}{\pi^2}\frac{mn}{(m^2+n^2)^2} \qquad \text{pour $m\neq n$ et $m+n$ est impair}$$
$$\int_0^Lx\sin\frac{m\pi x}{L}\sin\frac{n\pi x}{L}\text{d}x = \frac{L^2}{4} \qquad \text{pour $m=n$}$$
On peut donc résoudre les intégralespout les différents cas. Pour $m=n$, on a
$$H_{mn} = \frac{\hbar^2\pi^2n^2}{ML^3}\bigg(\frac{L}{2}\bigg) + \frac{2a}{L^2}\bigg(\frac{L^2}{4}\bigg)$$
$$H_{mn} = \frac{\hbar^2\pi^2n^2}{2ML^2} + \frac{a}{2}$$
Pour $m\neq n$ avec $m+n$ impair, on a
$$H_{mn} = \frac{2a}{L^2}\bigg(-\frac{4L^2}{\pi^2}\frac{mn}{(m^2+n^2)^2}\bigg)$$
$$H_{mn} = -\frac{8amn}{\pi^2(m^2+n^2)^2}$$
Pour $m\neq n$ et $m+n$ pair, on a
$$H_{mn} = 0$$

La matrice $\boldsymbol{H}$ a donc la forme suivante :
$$\begin{bmatrix}
H_{nn} & H_{mn} & 0 & H_{mn} & 0 & \ldots\\
H_{mn} & H_{nn} & H_{mn} & 0 & H_{mn} & \ldots\\
0 & H_{mn} & H_{nn} & H_{mn} & 0 & \ldots\\
H_{mn} & 0 & H_{mn} & H_{nn} & H_{mn} & \ldots\\
0 & H_{mn} & 0 & H_{mn} & H_{nn} & \ldots\\
\vdots & \vdots & \vdots & \vdots & \vdots & \ddots
\end{bmatrix}$$
où
$$H_{nn} = \frac{\hbar^2\pi^2n^2}{2ML^2} + \frac{a}{2}$$
$$H_{mn} = -\frac{8amn}{\pi^2(m^2-n^2)^2} = -\frac{8anm}{\pi^2(n^2-m^2)^2} = H_{nm}$$

On observe donc que la matrice est réelle et symétrique. On code ensuite la fonction *element_H* qui calcule l'élément $m$, $n$ de la matrice $\boldsymbol{H}$ en prenant les valeurs de $m$, $n$, $a$, $L$ et $M$. Tout d'abord, le code compare les valeurs de $m$ et $n$ et détermine la parité de $m+n$ afinde déterminer la formule qui s'applique. Elle calcule ensuite la valeur de l'élément.

In [77]:
# Fonction qui calcule les éléments de la matrice H
# Arguments:
# m, n : indices de l'élément
# a : paramètre a [eV]
# L : largeur du puit de potentiel [Angstrom]
# M : masse de la particule [kg]
# Retourne l'élément de matrice [eV]

def element_H(m, n, a, L, M):
    a = a / (6.241509 * 10**18)
    L = L * 10**(-9)
    if m == n:
        return (hbar**2*pi**2*n**2 / (2*M*L**2) + a / 2) * 6.241509 * 10**18
    elif (m+n) % 2 == 0:
        return 0.
    else:
        return -8*a*m*n / (pi**2*(m**2-n**2)**2) * 6.241509 * 10**18

On crée maintenant *element_H_mn*, une fonction spécifique aux données du problème, qui sont $a=10$ eV et $L=0,5$ nm avec $M$ qui correspond à la masse de l'électron. Cette fonction prend uniquement $m$ et $n$ en argument.

In [78]:
def element_H_mn(m, n):
    return element_H(m, n, 10, 0.5, me)

## Question 6

La matrice $H$ est en théorie infinie, donc il est évidemment impossible de calculer tous ses éléments. On va donc devoir approximer que la matrice $H$ a des dimensions finies. On définit la fonction *matrice_H* qui génère un *array numpy* de la matrice $H$ de dimensions $N$ par $N$. Cette fonction construit d'abord un *array* vide de la bonne dimension et calcule ensuite chaque élément à l'aide de la fonction *element_H_mn*.

In [79]:
# Construit la matrice H à partir de la fonction elements_H_mn
# Argument:
# N : dimension de la matrice (NxN)
# Retourne un array numpy de dimensions NxN représentant la matrice H

def matrice_H(N):
    H = np.zeros((N, N))
    for m in range(N):
        for n in range(N):
            H[m,n] = element_H_mn(m+1, n+1)
    return H

En première approximation, on détermine la matrice $H$ pour $N=10$. On veut ensuite déterminer ses valeurs propres, qui correspondent aux 10 premiers niveaux d'énergie. Pour ce faire, on utilise la fonction *eigvalsh* de la librairie *numpy.linalg* [1]. Cette fonction utilise la décomposition QR de façon itérative afin de déterminer les valeurs propres d'une matrice hermitienne. La décomposition QR consiste à représenter une matrice $X$ comme le produit d'une matrice orthogonale $Q$ et d'une matrice triangulaire supérieure $R$. Pour déterminer les valeurs propres de la matrice, on décompose la matrice $X_k=Q_kR_k$ de la $k$-ième itération pour ensuite définir la prochaine itération comme $X_{k+1}=R_kQ_k$. Au fil des itérations, la matrice $X_k$ va tendre vers une matrice diagonale dont les éléments de la diagonales principale sont les valeurs propres recherchées. La fonction *eigvalsh* a été choisie entre autres pour son efficacité, car le fait que la matrice donnée en argument soit hermitienne permet à la fonction d'utiliser les propriétés des matrices hermitiennes pour réduire le temps de calcul. Dans notre cas, comme la matrice $H$ est symétrique et réelle, on sait qu'elle est hermitienne.

In [81]:
# On crée la matrice H 10x10
H_10 = matrice_H(10)

#print(H_10)

# On trouve les valeurs propres avec la fonction de numpy
valpropres_10 = np.linalg.eigvalsh(H_10).tolist()
valpropres_10.sort()
print('Les valeurs propres de la matrice H 10x10 sont :', valpropres_10)

Les valeurs propres de la matrice H 10x10 sont : [5.84201832068847, 11.19716835721644, 18.69923015075477, 29.20917138991124, 42.75678500075369, 59.3318232668801, 78.92891366542653, 101.54616309119842, 127.18133344813275, 155.96234690491175]


On a donc les 10 niveau d'énergie suivants pour le puit de potentiel :

<center>

| Niveau     | Énergie<br />[eV] |
|:------------:|:--------------:|
| 1          | 5.8420183      |
| 2          | 11.197168      |
| 3          | 18.699230      |
| 4          | 29.209171      |
| 5          | 42.756785      |
| 6          | 59.331823      |
| 7          | 78.928914      |
| 8          | 101.54616      |
| 9          | 127.18133      |
| 10         | 155.96235      |

<center>
<center> Tableau N. Niveaux d'énergie pour une matrice 10x10 <center>

## Question 7

On veut maintenant améliorer notre modèle en considérant une matrice de dimensions 100x100. On construit donc la matrice correspondante et on trouve les valeurs propres de la même façon qu'à la question 6.

In [83]:
# On crée la matrice H 100x100
H_100 = matrice_H(100)

#print(H_10)

# On trouve les valeurs propres avec la fonction de numpy
valpropres_100 = np.linalg.eigvalsh(H_100).tolist()
valpropres_100.sort()
print('Les valeurs propres de la matrice H 100x100 sont :', valpropres_100)

Les valeurs propres de la matrice H 100x100 sont : [5.8420179222420785, 11.197167037678128, 18.699228285281336, 29.209162614796522, 42.75677590734573, 59.33177083204045, 78.92886199851962, 101.54553399316707, 127.18050485095517, 155.83307295064813, 187.50283009621603, 222.18952636543773, 259.89300205613245, 300.61315136087484, 344.34990198050883, 391.10320317811215, 440.8730185122354, 493.65932126996574, 549.4620915132483, 608.2813141195106, 670.1169774521167, 734.9690724397315, 802.8375919270802, 873.7225302094504, 947.6238826938844, 1024.5416456491632, 1104.4758160189301, 1187.4263912804736, 1273.3933693367223, 1362.3767484329833, 1454.3765270919685, 1549.3927040628603, 1647.4252782808699, 1748.4742488350555, 1852.5396149423548, 1959.6213759267023, 2069.7195312018916, 2182.834080257698, 2298.9650226482695, 2418.112357982732, 2540.2760859171317, 2665.456206147986, 2793.652718406519, 2924.865622454201, 3059.0949180785697, 3196.3406050900694, 3336.6026833189458, 3479.8811526130685, 3626

On regroupe les 10 premiers niveaux d'énergie pour les matrices 10x10 et 100x100 afin de comparer les valeurs obtenues :


<center>

| Niveau     | Énergie - 10x10<br />[eV]  | Énergie - 100x100<br />[eV]  | 
|:------------:|:--------------:|:--------------:|
| 1          | 5.8420183      | 5.8420179      |
| 2          | 11.197168      | 11.197167      |
| 3          | 18.699230      | 18.699228      |
| 4          | 29.209171      | 29.209163      |
| 5          | 42.756785      | 42.756776      |
| 6          | 59.331823      | 59.331771      |
| 7          | 78.928914      | 78.928862      |
| 8          | 101.54616      | 101.54553      |
| 9          | 127.18133      | 127.18050      |
| 10         | 155.96235      | 155.83307      |

<center>
<center> Tableau N. Niveaux d'énergie pour les matrices 10x10 et 100x100 <center>

On observe que les valeurs d'énergie de la matrice 10x10 et de la matrice 100x100 sont très rapprochées. La différence la plus importante est pour le niveau 10 et est l'ordre de $10^{-4}$ fois la valeur du niveau d'énergie. On remarque également que l'erreur relative remble augmenter avec les niveaux d'énergie.

## Médiagraphie

[1] (eigenvalues) https://numpy.org/doc/stable/reference/generated/numpy.linalg.eigvals.html